source/common/statistic_impl.h

```diff
@@ -150,4 +151,118 @@ class HdrStatistic : public StatisticImpl {
   struct hdr_histogram* histogram_;
 };
 
-} // namespace Nighthawk
\ No newline at end of file
+/**
+ * CircllhistStatistic uses Circllhist under the hood to compute statistics.
+ * Circllhist is used in the implementation of Envoy Histograms, compared to HdrHistogram it trades
+ * precision for fast performance in merge and insertion. For more info, please see
+ * https://github.com/circonus-labs/libcircllhist
+ */
+class CircllhistStatistic : public StatisticImpl {
+public:
+  CircllhistStatistic() {
+    histogram_ = hist_alloc();
+    ASSERT(histogram_ != nullptr);
+  }
+  ~CircllhistStatistic() override { hist_free(histogram_); }
+
+  void addValue(uint64_t value) override {
+    hist_insert_intscale(histogram_, value, 0, 1);
+    StatisticImpl::addValue(value);
+  }
+  double mean() const override { return hist_approx_mean(histogram_); }
+  double pvariance() const override { return pstdev() * pstdev(); }
+  double pstdev() const override {
+    return count() == 0 ? std::nan("") : hist_approx_stddev(histogram_);
+  }
+  StatisticPtr combine(const Statistic& statistic) const override;
+  uint64_t significantDigits() const override { return 1; }
+  StatisticPtr createNewInstanceOfSameType() const override {
+    return std::make_unique<CircllhistStatistic>();
+  }
+  nighthawk::client::Statistic toProto(SerializationDomain domain) const override;
+
+private:
+  histogram_t* histogram_;
```

Comment:
(optional) Using a raw pointer here opens us to memory leaks and all the usual problems of raw pointers.

Would it be possible to move this to a smart pointer (std::unique_ptr) with a custom deleter?

https://www.bfilipek.com/2016/04/custom-deleters-for-c-smart-pointers.html

Reply:
Thanks Jakub for the reference. It seems the [`hist_free(histogram_t *hist)`](https://github.com/circonus-labs/libcircllhist/blob/18f1b97603757d5e65baefa65b90d6ec182e8a30/src/circllhist.c#L1447) is quite complicated, so I will not move it to a smart pointer for now